### ISSUE TITLE
pylint not finding members on trio library

### ISSUE THREAD

**Issue by jmfrank63** [PyCQA/pylint#2382]
### Steps to reproduce
Use this code:
```
# -*- coding: utf-8 -*-

import trio

async def toddler():
    print('Toddler is playing...')
    await trio.sleep(5)

async def parent():
    async with trio.open_nursery() as nursery: #pylint: disable=no-member
        print('Bringing toddler to nursery...')
        nursery.start_soon(toddler)
    print('Going home...')

if __name__ == '__main__':
    trio.run(parent)

```

In this code above pylint on python 3.7 complains about not finding members:

```

### Current behavior

##########Linting Output - pylint##########
************* Module issue
10,15,error,E1101:Module 'trio' has no 'open_nursery' member
17,4,error,E1101:Module 'trio' has no 'run' member

------------------------------------------------------------------
Your code has been rated at 1.67/10 (previous run: 1.67/10, +0.00)
```
The issue can obviously worked around by ignoring.

### Expected behavior
Pylint not reporting any errors.

### pylint --version output
pylint 2.1.0
astroid 2.0.2
Python 3.7.0 (default, Jul 29 2018, 18:10:54)
[Clang 9.1.0 (clang-902.0.39.2)]

on MacBook Air OSX 10.12

ipython finds both functions with code completion.

Issue is also reported against trio.

**Comment by jmfrank63** [comment]
I checked out the way trio exposes its methods. It starts with an empty `__all__` list and then imports modules adding their methods dynamically to the list via `+=`
I don't currently know how difficult it might be for pylint to follow the imports and importing all the different `__all__` lists to combine them in one, and doing this possibly even over multiple instances.
Any comments?

**Comment by PCManticore** [comment]
@jmfrank63 Thanks for creating an issue!

So `pylint` stopped looking a couple of years ago at `__all__`. The problem was like this example, folks tended to modify `__all__` dynamically which resulted in false positives in `pylint` due to it not being able to follow all the paths for inferring what potential values might be added to that list. With that being said, `pylint` is still able to find all your members, just that it doesn't restrict itself to finding just those from `__all__`. 

In `trio`'s example the problem is not the dynamically built `__all__`, but the fact that `run` is added using `globals()` which is something that `pylint` cannot make sense of: https://github.com/python-trio/trio/blob/master/trio/_toplevel_core_reexports.py#L34

The way we usually solve these issues is by adding a `brain` tip to astroid, pylint's inference engine: https://github.com/PyCQA/astroid/tree/master/astroid/brain. `brain` tips are just a form of patching an AST so that pylint can be able to find a member or another. Here's a simple one for example, where we just build an AST for the `collections` module and return it: https://github.com/PyCQA/astroid/blob/master/astroid/brain/brain_collections.py#L67

With that being said, the best persons to add those `brain` tips is for the folks actually using that library :) So if you have some time, would you mind sending a rough patch that adds `trio` there?

**Comment by njsmith** [comment]
FYI, we plan to fix this on the trio side instead, since that's easier than fixing all the different static analysis engines :-). See: https://github.com/python-trio/trio/pull/594

Thank you very much for the details on how pylint's inference actually works! Figuring out these kinds of details is definitely the hardest part of fixing this.
